modules/create_timepoint/php/timepoint.class.inc

```diff
@@ -0,0 +1,346 @@
+            $values['error'] = 'BadRequest';
+            return $values;
+        }
+        $values['identifier'] = isset($values['identifier']) ?
+            intval($values['identifier']) : null;
+        $values['psc']        = empty($values['psc']) ?
+            null : $values['psc'];
+
+        // validate site entered
+        $site = (int) $values['psc'];
+        $user_list_of_sites = $user->getData('CenterIDs');
+        $num_sites          = count($user_list_of_sites);
+        $conflict           = [];
+        if ($num_sites == 0) {
+            $conflict['psc']
+                = "You must be affiliated with a site to create a timepoint.";
+        } else if ($num_sites > 1 && (empty($site) || !$user->hasCenter($site))) {
+            $conflict['psc']
+                = "Site must be selected from the available dropdown.";
+        }
+
+        //Project
+        $user_list_of_projects = $user->getProjectIDs();
+        if (count($user_list_of_projects) == 0) {
+            $conflict['project']
+                = "You must be affiliated with a project to create a timepoint.";
+        } else if (count($user_list_of_projects) > 1
+            && is_null($values['project'])
+        ) {
+            $conflict['project']
+                = "A project must be selected from the available dropdown.";
+        }
+
+        $candid       = $values['candID'];
+        $subprojectID = intval($values['subproject']);
+        $visitLabel   = $values['visit'] ?? '';
+
+        try {
+            \TimePoint::isValidVisitLabel(
+                new CandID($candid),
+                $subprojectID,
+                $visitLabel
+            );
+        } catch (\LorisException $exception) {
+            $conflict['visitLabel'] = $exception->getMessage();
+        }
+
+        // validate language
+        $languages = \Utility::getLanguageList();
+        if (count($languages) > 1 && empty($values['languages'])) {
+            $conflict['languages'] = 'A language must be selected.';
+        }
+        // set form message if exists.
+        if ($conflict) {
+            $values['conflict'] = $conflict;
+        }
+        return $values;
+    }
+
+    /**
+     * Returns true if the user has permission to access
+     * the Create Timepoint module
+     *
+     * @param \User $user The user whose access is being checked
+     *
+     * @return bool true if user has permission
+     */
+    function _hasAccess(\User $user) : bool
+    {
+        return $user->hasPermission('data_entry');
+    }
+    /**
+     * An ETagCalculator provides the ability to calculate an ETag for
+     * an incoming HTTP request.
+     *
+     * @param ServerRequestInterface $request The incoming PSR7 request.
+     *
+     * @return string The value to use for the ETag header.
+     */
+    public function ETag(ServerRequestInterface $request): string
+    {
+        if ($request->getMethod() === 'POST') {
+            return 'Etag not calculated on POST request';
```

Comment:
This will set an ETag of "Etag not calculated on POST request". It should be an empty string to avoid adding an ETag.

Reply:
thanks @driusan, see latest commits